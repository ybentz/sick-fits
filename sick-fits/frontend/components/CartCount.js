import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { TransitionGroup, CSSTransition } from 'react-transition-group';

const AnimationStyles = styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: transform 0.3s;
    backface-visibility: hidden;
  }
  .count-enter {
    transform: rotateX(0.5turn);
  }
  .count-enter-active {
    transform: rotateX(0);
  }
  .count-exit {
    transform: rotateX(0);
    top: 0;
    position: absolute;
  }
  .count-exit-active {
    transform: rotateX(0.5turn);
  }
`;

const Dot = styled.div`
  background: ${props => props.theme.red};
  color: white;
  border-radius: 50%;
  padding: 0.25rem;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-weight: 100;
  /* keep the dot the same size regardless of how wide the number inside is */
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
`;

const CartCount = ({ count }) => (
  <AnimationStyles>
    <TransitionGroup>
      <CSSTransition
        unmountOnExit
        className="count"
        classNames="count"
        key={count}
        timeout={{ enter: 300, exit: 300 }}
      >
        <Dot>{count}</Dot>
      </CSSTransition>
    </TransitionGroup>
  </AnimationStyles>
);

export default CartCount;
