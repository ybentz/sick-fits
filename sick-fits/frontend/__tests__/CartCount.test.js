import '@testing-library/jest-dom'
import { render, waitFor } from '@testing-library/react'
import React from 'react'

import CartCount from '../components/CartCount'

describe('<CartCount/>', () => {
  it('updates via props', async () => {
    const count = 50
    const { asFragment, container, rerender } = render(
      <CartCount count={count} />
    )
    expect(container).toHaveTextContent(count)
    // expect(asFragment()).toMatchSnapshot()
    const updatedCount = 10
    rerender(<CartCount count={updatedCount} />)
    // Wait for css transition to finish - there's probably a better way to do this
    // Should check this out https://testing-library.com/docs/example-react-transition-group
    await waitFor(() => {
      // There seems to be a better way to wait for element to disappear but I'm not sure how
      // it works with the container https://testing-library.com/docs/guide-disappearance#waiting-for-disappearance
      expect(container).not.toHaveTextContent(count)
    })
    expect(container).toHaveTextContent(updatedCount)
    // expect(asFragment()).toMatchSnapshot()
  })
})
