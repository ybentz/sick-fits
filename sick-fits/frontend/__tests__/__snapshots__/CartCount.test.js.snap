// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartCount/> updates via props 1`] = `
<DocumentFragment>
  <span
    class="CartCount__AnimationStyles-xxvp4g-0 jGgtAN"
  >
    <div>
      <div
        class="CartCount__Dot-xxvp4g-1 kCeHkh count"
      >
        50
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`<CartCount/> updates via props 2`] = `
<DocumentFragment>
  <span
    class="CartCount__AnimationStyles-xxvp4g-0 jGgtAN"
  >
    <div>
      <div
        class="CartCount__Dot-xxvp4g-1 kCeHkh count count-enter-done"
      >
        10
      </div>
    </div>
  </span>
</DocumentFragment>
`;
